### ISSUE TITLE
Found some mathematical problems that seem to be wrong about the lighthouse calibration value 

### ISSUE THREAD

**Issue by madzhen** [bitcraze/crazyflie-firmware#1423]
I don't think this question is important right now, I will close it

**Comment by knmcguire** [comment]
Hi!

There are some assumptions here that we'd need to clear out:

> I also found the correct formula in this link, which uses a uniform minus sign:
> 
> https://github.com/cntools/libsurvive/wiki/BSD-Calibration-Values#phase
> 

You mention that this is the correct formula since this is from the Libsurvive, however, they are also trying to approximate the calibration model just as we are. I quote from the link that you have provided above in their words:

> Obviously, for perfect calibration we'd want to know and use the vives exact formulation -- down to their choice of signs, offsets, scales, etc.
> 
> If our goal though is just to get arbitrarily close, we can approximate these formulations. The first thing we do is add a scalar term to each parameter.

The actual calibration model is unknown to both Libsurvive and us at Bitcraze, and we have made efforts to try to correct for it as much as wel can. The reason why we use different models and formula's is that Libsurvive 's library is used for virtual reality applications, and we use it for onboard UAV positioning. Both of these have different requirements on how accurate the systems has to be.

So long story short, there is no right formula since Vive/SteamVR hasn't released that to us and both us and Libsurvive have tried to approximate and interpret it that the best we can, with the calibration values we got from the basestations and the requirements in which we use the Lighthouse system for. There is no right or wrong formula here, just different approaches.

Please don't directly attach peoples handles in the github discussion posts. We try to answer the github discussions to the best of our abilities and we get a notice whenever there is a new post. We are just a small company and we try to answer all as quickly as we can with the time that we have next to development.

**Comment by madzhen** [comment]
> Hi!
> 
> There are some assumptions here that we'd need to clear out:
> 
> > I also found the correct formula in this link, which uses a uniform minus sign:
> > https://github.com/cntools/libsurvive/wiki/BSD-Calibration-Values#phase
> 
> You mention that this is the correct formula since this is from the Libsurvive, however, they are also trying to approximate the calibration model just as we are. I quote from the link that you have provided above in their words:
> 
> > Obviously, for perfect calibration we'd want to know and use the vives exact formulation -- down to their choice of signs, offsets, scales, etc.
> > If our goal though is just to get arbitrarily close, we can approximate these formulations. The first thing we do is add a scalar term to each parameter.
> 
> The actual calibration model is unknown to both Libsurvive and us at Bitcraze, and we have made efforts to try to correct for it as much as wel can. The reason why we use different models and formula's is that Libsurvive 's library is used for virtual reality applications, and we use it for onboard UAV positioning. Both of these have different requirements on how accurate the systems has to be.
> 
> So long story short, there is no right formula since Vive/SteamVR hasn't released that to us and both us and Libsurvive have tried to approximate and interpret it that the best we can, with the calibration values we got from the basestations and the requirements in which we use the Lighthouse system for. There is no right or wrong formula here, just different approaches.
> 
> Please don't directly attach peoples handles in the github discussion posts. We try to answer the github discussions to the best of our abilities and we get a notice whenever there is a new post. We are just a small company and we try to answer all as quickly as we can with the time that we have next to development.

I completely understand what you are saying. I also completely understand the actual meaning of the mathematical formulas used by libsurvive and you. I know very well that you are using the same method. I said that the negative sign here may be reversed because I have derived the formulas here in detail. Therefore, I hope you can read the details of the problem I described above carefully. I have spent about 4 months on the derivation of the mathematical formulas here, so I know very well the specific meanings of all the formulas you use.

Both libsurvive and you give an approximate method. Regarding gibphase, they all give the first-order Fourier transform terms. I have read all the codes in detail. The core code used by libsurvive and you are exactly the same, and the writing method is exactly the same. Then you have an extra negative sign.

I also carefully checked the astrobee project. Your calculation formula comes from astrobee, because Kristoffer's previous comments detailed the source of this part of the code, and there is no negative sign in the astrobee code. The reason I raised this question is that I have derived all the mathematical processes in detail, and I can be sure that the negative sign here is wrong. Because whether it is libsurvive or astrobee, the raw data sent by the base station is exactly the same as yours.

As far as I know, astrobee and libsurvive both have high accuracy, so I said that the negative sign is likely the culprit. I was on a business trip recently so I couldn't actually test it.

**Comment by Bananazooka** [comment]
Hi, we appreciate all the work you put into this; just a friendly reminder to keep things respectful as per Github [discussion rules](https://github.com/orgs/bitcraze/discussions/55). We're all here to help each other!
As for the firmware, it's open source, so you're welcome to propose a fix when you've flight-tested it.
